### PR TITLE
Listeners on Parser.map()

### PR DESCRIPTION
**Pull request by scolomer**
Hello,

I implemented a way to add runtime listeners to a parse. The listener il called when Parse.map() is called. For instance, it permits the implementation of "Locatable", as we discussed some time ago, in a non-intrusive way. 
I call it Parameters because I see it as a value holder for some other features. I intend to use I soon for the implementation of some conditionnal parsing.

### CODE REVIEW

**Comment by fluentfuture** [comment]
Thanks Arnaud!

Yes. I'd like to have a discussion on the high level use case and high level API, before diving into impl or tests.

Off the top of my head, these are the kind of things I think worth discussing:
* The use case.
  * What problem is it targeted to solve?
  * Could it be implemented with existing API?
  * If doable with existing API, is it a documentation bug? discoverability bug? Is it used often enough to justify a dedicated convenience API?
* The proposed API.
  * Will it work? Will it play nicely with other combinators? Or is there some subtle caveats?
  * Is it the level of abstraction we want to attack the problem at? Can it be more generic? more specific?
  * What do we want the API look like? Should it live in Parser.java, Parsers.java or Scanners.java? What's the method signature etc.

**Comment by scolomer** [comment]
Hi Ben, Hi Arnaud,

No problem to have a discussion on this feature, I developped this to fit my own needs but I issued the pull request because I don't want to maintain my own fork of JParsec and I would be happy to bring new features to other users.

Basically I use JParsec to build AST from programming languages. With the AST I can implement Eclipse text editor features like hyperlinks, tooltips, refactoring and other static analysis stuff.

Today I miss 2 features in JParsec:
* Have a generic way to populate "location" information on every node of my AST. By "location" I mean the position of the instruction in the input file that matched the parser.
* Enable/disable parsing feature at runtime. For instance if I write a Java parser I would probably want to enable or disable generics and enums depending on the version of java I want to parse.

The first point can be implemented with the existing API, I've done it with a slightly modified version of JParsec 2.0.1. I know I could have achieved this with Parsers.INDEX with no modification. The drawback is I have to modify quite every rule of my parser to add a ".location()" of enclose it into "sourceInfo()" which is very tedious and error prone. Considering previous feature request, I'm pretty sure I'm not the only one to have such needs. We already had discussions on this point (https://github.com/jparsec/jparsec/issues/5, https://github.com/jparsec/jparsec/pull/7), Parser.withSource() that was added to the API (https://github.com/jparsec/jparsec/pull/16) is somehome related to this point.

To my point of view there is no "clean" way to implement the second point. I've no idea if I'm the only one to have such a need.

About the implementation of the first point, I'm not sure this has to be at Parser.parse level. It also could be defined at parser construction level : parser.from(...).withListener(new Listener(){}). 
Basically, I wanted to implement this feature in Parser.apply(), but there are many overrides of this method and it's not easy to correlate Parser.apply() and ParserContext.result. I think Parser.map() is quite a good place because it is usefull when a new instance is created.

I also like the idea of generic parameters that can be passed at runtime to the Parser.parse() method to be used in Parsers, Listeners...

I did not thought a lot about the implementation of the second point but I guess it could be something like Parser.condition(new Condition() {...}).

**Comment by fluentfuture** [comment]
Agreed that it shouldn't be on the parse() level. This is the API I can imagine:

```
parser.withListener(Listener);
```
And the listener could be:
```
interface Listener {
  void onResult(Object result);
}
```

Listener can be nested, that is, if you call ```isChar('c').withListener(a).many().withListener(b)```, the char result should be passed to listener ```a``` and the List result should be intercepted by ```b```.

For implementation, all the parser results are stored in ```ParseContext#result``` field. There are a handful of places where result is directly read/written.

I think it's easy to do an "Encapsulate Field" refactoring by exposing a getResult() and setResult() method. And then in setResult() method, we call the listener object, is one is present.

The listener can be set into the ParseContext like this:

```
public Parser<T> withListener(final Listener listener) {
  checkNotNull(listener);
  return new Parser<T>() {
    Listener oldListener = ctxt.getListener();
    ctxt.setListener(listener);
    try {
      return Parser.this.apply(ctxt);
    } finally {
      ctxt.setListener(oldListener);
    }
  };
}
```

Regarding conditional grammar, could it be done by parameterizing the parser? For example, if your Java grammar is defined as:

```
Parser<AST> javaParser(Parser<TypeDefinition> typeDef, Parser<TypeRef> typeRef);
```
Then you can pass type parser with or without generics.

In other words, I wonder if this is an orthogonal problem solvable by parameterizing the parser code.

**Comment by abailly** [comment]
What do we do with this PR? Anyone still working on it?
